discovery/hetzner/hcloud.go

```diff
@@ -0,0 +1,125 @@
+// Copyright 2020 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package hetzner
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/hetznercloud/hcloud-go/hcloud"
+	config_util "github.com/prometheus/common/config"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/common/version"
+	"github.com/prometheus/prometheus/discovery/refresh"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+)
+
+const (
+	hetznerHcloudLabelPrefix           = hetznerLabelPrefix + "hcloud_"
+	hetznerLabelHCloudImageName        = hetznerHcloudLabelPrefix + "image_name"
+	hetznerLabelHCloudImageDescription = hetznerHcloudLabelPrefix + "image_description"
+	hetznerLabelHCloudPrivateIPv4      = hetznerHcloudLabelPrefix + "private_ipv4"
+	hetznerLabelHCloudNetwork          = hetznerHcloudLabelPrefix + "network"
+	hetznerLabelHCloudLocation         = hetznerHcloudLabelPrefix + "location"
+	hetznerLabelHCloudCPUCores         = hetznerHcloudLabelPrefix + "cpu_cores"
+	hetznerLabelHCloudCPUType          = hetznerHcloudLabelPrefix + "cpu_type"
+	hetznerLabelHCloudMemoryGB         = hetznerHcloudLabelPrefix + "memory_size_gb"
+	hetznerLabelHCloudDiskGB           = hetznerHcloudLabelPrefix + "disk_size_gb"
+	hetznerLabelHCloudType             = hetznerHcloudLabelPrefix + "server_type"
+)
+
+// Discovery periodically performs Hetzner Cloud requests. It implements
+// the Discoverer interface.
+type HcloudDiscovery struct {
+	*refresh.Discovery
+	client  *hcloud.Client
+	port    int
+	network *hcloud.Network
+}
+
+// NewDiscovery returns a new Discovery which periodically refreshes its targets.
+func newHCloudDiscovery(conf *SDConfig, logger log.Logger) (*HcloudDiscovery, error) {
+	d := &HcloudDiscovery{
+		port: conf.Port,
+	}
+
+	rt, err := config_util.NewRoundTripperFromConfig(conf.HTTPClientConfig, "hetzner_sd", false)
+	if err != nil {
+		return nil, err
+	}
+	opts := []hcloud.ClientOption{
+		hcloud.WithApplication("Prometheus", version.Version),
+		hcloud.WithHTTPClient(&http.Client{
+			Transport: rt,
+			Timeout:   time.Duration(conf.RefreshInterval),
+		}),
+		hcloud.WithPollInterval(time.Duration(conf.RefreshInterval)),
+		hcloud.WithEndpoint(conf.HCloudEndpoint),
```

Comment:
should HCloudEndpoint be public? Why would users override it? We could make it private to limit it to tests.